# Python String `format()`

The string **`format()`** method formats the given string into a nicer output in Python.

**Syntax**:

```python
template.format(p0, p1, ..., k0=v0, k1=v1, ...)
```

Here, **`p0, p1,...`** are positional arguments and, **`k0, k1,...`** are keyword arguments with values **`v0, v1,...`** respectively.

And, **`template`** is a mixture of format codes with placeholders for the arguments


## `format()` Parameters

The **`format()`** method takes any number of parameters. But, is divided into two types of parameters:

- **Positional parameters** - list of parameters that can be accessed with index of parameter inside curly braces **`{index}`**
- **Keyword parameters** - list of parameters of type key=value, that can be accessed with key of parameter inside curly braces **`{key}`**


## Return Value from `format()`

The **`format()`** method returns the formatted string.


## How String `format()` works?

The **`format()`** reads the type of arguments passed to it and formats it according to the format codes defined in the string.


### For positional arguments

<div>
<img src="img/format1.png" width="600"/>
</div>


Here, Argument **0** is a string **"Adam"** and Argument **1** is a floating number **230.2346**.

> **Note:** Argument list starts from **0** in Python.

The string **`"Hello {0}, your balance is {1:9.3f}"`** is the template string. This contains the format codes for formatting.

The curly braces are just placeholders for the arguments to be placed. In the above example, **`{0}`** is placeholder for **`"Adam"`** and **`{1:9.3f}`** is placeholder for **`230.2346`**.

Since the template string references **`format()`** arguments as **`{0}`** and **`{1}`**, the arguments are positional arguments. They both can also be referenced without the numbers as **`{}`** and Python internally converts them to numbers.

Internally,

- Since **`"Adam"`** is the 0th argument, it is placed in place of **`{0}`**. Since, **`{0}`** doesn't contain any other format codes, it doesn't perform any other operations.

- However, it is not the case for 1st argument **`230.2346`**. Here, **`{1:9.3f}`** places **`230.2346`** in its place and performs the operation **9.3f**.

- **f** specifies the format is dealing with a float number. If not correctly specified, it will give out an error.

- The part before the **"." (9)** specifies the minimum width/padding the number **(230.2346)** can take. In this case, **230.2346** is allotted a minimum of **9** places including the **"."**.

- If no alignment option is specified, it is aligned to the right of the remaining spaces. (For strings, it is aligned to the left.)

- The part after the **"." (3)** truncates the decimal part **(2346)** upto the given number. In this case, **2346** is truncated after 3 places.

- Remaining numbers **(46)** is rounded off outputting **235**.


### For keyword arguments

<div>
<img src="img/format2.png" width="600"/>
</div>


We've used the same example from above to show the difference between keyword and positional arguments.

Here, instead of just the parameters, we've used a key-value for the parameters. Namely, **`name="Adam"`** and **`blc=230.2346`**.

Since these parameters are referenced by their keys as **`{name}`** and **`{blc:9.3f}`**, they are known as keyword or named arguments.

Internally,

- The placeholder **`{name}`** is replaced by the value of name - "Adam". Since it doesn't contain any other format codes, "Adam" is placed.

- For the argument **blc=230.2346**, the placeholder **`{blc:9.3f}`** is replaced by the value 230.2346. But before replacing it, like previous example, it performs 9.3f operation on it.

This outputs **230.235**. The decimal part is truncated after **3** places and remaining digits are rounded off. Likewise, the total width is assigned 9 leaving two spaces to the left.


## Basic formatting with `format()`

The **`format()`** method allows the use of simple placeholders for formatting.


In [ ]:
# Example 1: Basic formatting for default, positional and keyword arguments

# default arguments
print("Hello {}, your balance is {}.".format("Adam", 230.2346))

# positional arguments
print("Hello {0}, your balance is {1}.".format("Adam", 230.2346))

# keyword arguments
print("Hello {name}, your balance is {blc}.".format(name="Adam", blc=230.2346))

# mixed arguments
print("Hello {0}, your balance is {blc}.".format("Adam", blc=230.2346))

SyntaxError: positional argument follows keyword argument (3282176356.py, line 13)

> **Note:** In case of mixed arguments, keyword arguments has to always follow positional arguments.


## Numbers formatting with `format()`

You can format numbers using the format specifier given below:

| Type    | Meaning                                                                           |
| :------ | :-------------------------------------------------------------------------------- |
| **`d`** | **Decimal integer**                                                               |
| **`c`** | **Corresponding Unicode character**                                               |
| **`b`** | **Binary format**                                                                 |
| **`o`** | **Octal format**                                                                  |
| **`x`** | **Hexadecimal format (lower case)**                                               |
| **`X`** | **Hexadecimal format (upper case)**                                               |
| **`n`** | **Same as 'd'. Except it uses current locale setting for number separator**       |
| **`e`** | **Exponential notation. (lowercase e)**                                           |
| **`E`** | **Exponential notation (uppercase E)**                                            |
| **`f`** | **Displays fixed point number (Default: 6)**                                      |
| **`F`** | **Same as 'f'. Except displays 'inf' as 'INF' and 'nan' as 'NAN'**                |
| **`g`** | **General format. Rounds number to p significant digits. (Default precision: 6)** |
| **`G`** | **Same as 'g'. Except switches to 'E' if the number is large.**                   |
| **`%`** | **Percentage. Multiples by 100 and puts % at the end.**                           |


In [2]:
# Example 2: Simple number formatting

# integer arguments
print("The number is:{:d}".format(123))

# float arguments
print("The float number is:{:f}".format(123.4567898))

# octal, binary and hexadecimal format
print("bin: {0:b}, oct: {0:o}, hex: {0:x}".format(12))

The number is:123
The float number is:123.456790
bin: 1100, oct: 14, hex: c


In [3]:
# Example 3: Number formatting with padding for int and floats
    
# integer numbers with minimum width
print("{:5d}".format(12))

# width doesn't work for numbers longer than padding
print("{:2d}".format(1234))

# padding for float numbers
print("{:8.3f}".format(12.2346))

# integer numbers with minimum width filled with zeros
print("{:05d}".format(12))

# padding for float numbers filled with zeros
print("{:08.3f}".format(12.2346))

   12
1234
  12.235
00012
0012.235


**Explanation:** Here,

- in the first statement, **`{:5d}`** takes an integer argument and assigns a minimum width of 5. Since, no alignment is specified, it is aligned to the right.

- In the second statement, you can see the width (2) is less than the number (1234), so it doesn't take any space to the left but also doesn't truncate the number.

- Unlike integers, floats has both integer and decimal parts. And, the mininum width defined to the number is for both parts as a whole including **"."**.

- In the third statement, **`{:8.3f}`** truncates the decimal part into 3 places rounding off the last 2 digits. And, the number, now 12.235, takes a width of 8 as a whole leaving 2 places to the left.

- If you want to fill the remaining places with zero, placing a zero before the format specifier does this. It works both for integers and floats: **`{:05d}`** and **`{:08.3f}`**.


In [4]:
# Example 4: Number formatting for signed numbers

# show the + sign
print("{:+f} {:+f}".format(12.23, -12.23))

# show the - sign only
print("{:-f} {:-f}".format(12.23, -12.23))

# show space for + sign
print("{: f} {: f}".format(12.23, -12.23))

+12.230000 -12.230000
12.230000 -12.230000
 12.230000 -12.230000


## Number formatting with alignment

The operators **`<`**, **`^`**, **`>`** and **`=`** are used for alignment when assigned a certain width to the numbers.

| Type    | Meaning                                                |
| :------ | :----------------------------------------------------- |
| **`<`** | **Left aligned to the remaining space**                |
| **`^`** | **Center aligned to the remaining space**              |
| **`>`** | **Right aligned to the remaining space**               |
| **`=`** | **Forces the signed (+) (-) to the leftmost position** |


In [5]:
# Example 5: Number formatting with left, right and center alignment

# integer numbers with right alignment
print("{:5d}".format(12))

# float numbers with center alignment
print("{:^10.3f}".format(12.2346))

# integer left alignment filled with zeros
print("{:<05d}".format(12))

# float numbers with center alignment
print("{:=8.3f}".format(-12.2346))

   12
  12.235  
12000
- 12.235


> **Note:** Left alignment filled with zeros for integer numbers can cause problems as the 3rd example which returns 12000, rather than 12.


## String formatting with `format()`

As numbers, string can be formatted in a similar way with **`format()`**.


In [6]:
# Example 6: String formatting with padding and alignment

# string padding with left alignment
print("{:5}".format("cat"))

# string padding with right alignment
print("{:>5}".format("cat"))

# string padding with center alignment
print("{:^5}".format("cat"))

# string padding with center alignment
# and '*' padding character
print("{:*^5}".format("cat"))

cat  
  cat
 cat 
*cat*


In [7]:
# Example 7: Truncating strings with format()

# truncating strings to 3 letters
print("{:.3}".format("caterpillar"))

# truncating strings to 3 letters
# and padding
print("{:5.3}".format("caterpillar"))

# truncating strings to 3 letters,
# padding and center alignment
print("{:^5.3}".format("caterpillar"))

cat
cat  
 cat 


## Formatting class and dictionary members using `format()`

Python internally uses **`getattr()`** for class members in the form ".age". And, it uses **`__getitem__()`** lookup for dictionary members in the form "[index]".


In [8]:
# Example 8: Formatting class members using format()

# define Person class
class Person:
    age = 23
    name = "Adam"

# format age
print("{p.name}'s age is: {p.age}".format(p=Person()))

Adam's age is: 23


**Explanation:** Here, Person **`object`** is passed as a keyword argument **`p`**.

Inside the template string, Person's **`name`** and **`age`** are accessed using **`.name`** and **`.age`** respectively.


In [9]:
# Example 9: Formatting dictionary members using format()

# define Person dictionary
person = {'age': 23, 'name': 'Adam'}

# format age
print("{p[name]}'s age is: {p[age]}".format(p=person))

Adam's age is: 23


**Explanation:** Similar to class, **`person`** dictionary is passed as a keyword argument **`p`**.

Inside the template string, person's **`name`** and **`age`** are accessed using **`[name]`** and **`[age]`** respectively.


There's an easier way to format dictionaries in Python using **`str.format(**mapping)`\*\*.


In [10]:
# define Person dictionary

person = {'age': 23, 'name': 'Adam'}

# format age
print("{name}'s age is: {age}".format(**person))

Adam's age is: 23


**`**`\*\* is a format parameter (minimum field width).


## Arguments as format codes using `format()`

You can also pass format codes like precision, alignment, fill character as positional or keyword arguments dynamically.


In [11]:
# Example 10: Dynamic formatting using format()

# dynamic string format template
string = "{:{fill}{align}{width}}"

# passing format codes as arguments
print(string.format('cat', fill='*', align='^', width=5))

# dynamic float format template
num = "{:{align}{width}.{precision}f}"

# passing format codes as arguments
print(num.format(123.236, align='<', width=8, precision=2))

*cat*
123.24  


**Explanation:** Here,

- In the first example, 'cat' is the positional argument is to be formatted. Likewise, **`fill='*'`**, **`align='^'`** and **`width=5`** are keyword arguments.

- In the template string, these keyword arguments are not retrieved as normal strings to be printed but as the actual format codes **`fill`**, **`align`** and **`width`**. The arguments replaces the corresponding named placeholders and the string **'cat'** is formatted accordingly.

- Likewise, in the second example, **123.236** is the positional argument and, align, width and precision are passed to the template string as format codes.


## Extra formatting options with `format()`

**`format()`** also supports type-specific formatting options like datetime's and complex number formatting.

**`format()`** internally calls **`__format__()`** for datetime, while **`format()`** accesses the attributes of the complex number.

You can easily override the **`__format__()`** method of any object for custom formatting.


In [12]:
# Example 11: Type-specific formatting with format() and overriding __format__() method

import datetime
# datetime formatting
date = datetime.datetime.now()
print("It's now: {:%Y/%m/%d %H:%M:%S}".format(date))

# complex number formatting
complexNumber = 1+2j
print("Real part: {0.real} and Imaginary part: {0.imag}".format(complexNumber))

# custom __format__() method
class Person:
    def __format__(self, format):
        if(format == 'age'):
            return '23'
        return 'None'

print("Adam's age is: {:age}".format(Person()))

It's now: 2021/06/10 23:36:57
Real part: 1.0 and Imaginary part: 2.0
Adam's age is: 23


**Explanation:** Here,

- **For datetime:**

  - Current datetime is passed as a positional argument to the **`format()`** method.
  - And, internally using **`__format__()`** method, **`format()`** accesses the year, month, day, hour, minutes and seconds.

- **For complex numbers:**

  - 1+2j is internally converted to a ComplexNumber object.
  - Then accessing its attributes **`real`** and **`imag`**, the number is formatted.

- **Overriding **format**():**
  - Like datetime, you can override your own **`__format__()`** method for custom formatting which returns age when accessed as **`{:age}`**


You can also use object's **`__str__()`** and **`__repr__()`** functionality with shorthand notations using **`format()`**.

Like **`__format__()`**, you can easily override object's **`__str__()`** and **`__repr_()`** methods.


In [13]:
# Example 12: __str()__ and __repr()__ shorthand !r and !s using format()

# __str__() and __repr__() shorthand !r and !s
print("Quotes: {0!r}, Without Quotes: {0!s}".format("cat"))

# __str__() and __repr__() implementation for class
class Person:
    def __str__(self):
        return "STR"
    def __repr__(self):
        return "REPR"

print("repr: {p!r}, str: {p!s}".format(p=Person()))

Quotes: 'cat', Without Quotes: cat
repr: REPR, str: STR


The `format()` method in Python is a powerful tool for formatting strings. It allows you to embed variables, values, and expressions into a string by replacing placeholders with actual data. This method is highly versatile and supports positional formatting, keyword formatting, and even custom formatting for numbers, dates, and more.

---

### Syntax

```python
string.format(*args, **kwargs)
```

### Parameters

1. **`*args`**: Positional arguments used to replace numbered placeholders in the string.
2. **`**kwargs`\*\*: Keyword arguments used to replace named placeholders in the string.

### Return Value

- Returns a formatted string with the placeholders replaced by the provided values.

---

### Placeholder Types

1. **Positional Placeholders**:
   - Use `{}` or `{n}` to denote placeholders, where `n` is the position index of the argument.
2. **Keyword Placeholders**:
   - Use `{name}` to denote placeholders that will be replaced using keyword arguments.
3. **Custom Formatting**:
   - Use format specifications inside `{}` to format numbers, dates, and other values (e.g., `{:.2f}` for two decimal places).

---

### Examples

#### Example 1: Basic Positional Formatting

```python
name = "Alice"
age = 30
result = "My name is {} and I am {} years old.".format(name, age)
print(result)
# Output: My name is Alice and I am 30 years old.
```

#### Example 2: Positional Indexing

```python
result = "I love {0} and {1}!".format("Python", "JavaScript")
print(result)
# Output: I love Python and JavaScript!
```

#### Example 3: Keyword Formatting

```python
result = "My name is {name} and I am {age} years old.".format(name="Bob", age=25)
print(result)
# Output: My name is Bob and I am 25 years old.
```

#### Example 4: Custom Number Formatting

```python
pi = 3.14159
result = "The value of pi to 2 decimal places is {:.2f}.".format(pi)
print(result)
# Output: The value of pi to 2 decimal places is 3.14.
```

#### Example 5: Alignment and Padding

```python
result = "Left: {:<10} | Center: {:^10} | Right: {:>10}".format("A", "B", "C")
print(result)
# Output:
# Left: A         | Center:     B     | Right:          C
```

#### Example 6: Combining Positional and Keyword Arguments

```python
result = "Hello {0}, your score is {score}".format("Charlie", score=95)
print(result)
# Output: Hello Charlie, your score is 95
```

#### Example 7: Nested Formatting

```python
result = "The product of {0} and {1} is {2:.2f}".format(5, 2, 5 * 2)
print(result)
# Output: The product of 5 and 2 is 10.00
```

---

### Advanced Customizations

1. **Fill and Alignment**:
   - Use `:<`, `:>`, and `:^` for left-align, right-align, and center-align respectively.
   - Add fill characters like `:0<`, `:*>`, etc.

```python
result = "Aligned: {:*>10}".format("Text")
print(result)
# Output: Aligned: ******Text
```

2. **Numeric Formats**:
   - `:d` for integers.
   - `:.2f` for floating-point numbers with 2 decimal places.
   - `:x` for hexadecimal representation.

```python
number = 255
result = "Decimal: {0}, Hexadecimal: {0:x}".format(number)
print(result)
# Output: Decimal: 255, Hexadecimal: ff
```

3. **Accessing Elements from Lists/Dictionaries**:
   - You can format values directly from lists or dictionaries.

```python
data = {"name": "Alice", "age": 30}
result = "Name: {name}, Age: {age}".format(**data)
print(result)
# Output: Name: Alice, Age: 30
```

---

### Summary

The `format()` method:

- Enables dynamic string creation.
- Offers fine-grained control over formatting.
- Is backward-compatible and preferred over older techniques like `%` formatting.


# Python String `format()` Method: Complete Guide

## What is the `format()` method?

The `format()` method is a powerful string formatting technique in Python that allows you to create formatted strings by replacing placeholders `{}` with values. It provides more flexibility than the older `%` formatting and is the predecessor to f-strings (Python 3.6+).

## Basic Syntax

```python
"string with {} placeholders".format(values)
```

### Simple Examples

```python
# Basic usage
print("Hello, {}".format("Alice"))  # Hello, Alice

# Multiple placeholders
print("{} + {} = {}".format(2, 3, 2+3))  # 2 + 3 = 5
```

## Positional Formatting

```python
# Positional arguments
print("{0} is {1} years old".format("Bob", 25))  # Bob is 25 years old

# Reusing positions
print("{0} {1} {0}".format("abra", "cad"))  # abra cad abra
```

## Named Parameters

```python
# Keyword arguments
print("{name} works as a {job}".format(name="Sarah", job="developer"))

# Dictionary unpacking
data = {'name': 'Mike', 'age': 30}
print("{name} is {age} years old".format(**data))
```

## Formatting Types

### Number Formatting

```python
# Integer formatting
print("{:d}".format(42))  # 42

# Float formatting
print("{:.2f}".format(3.14159))  # 3.14
print("{:10.2f}".format(3.14159))  # '      3.14' (10 width)

# Percentage
print("{:.1%}".format(0.75))  # 75.0%
```

### Alignment and Padding

```python
# Left align
print("{:<10}".format("left"))    # 'left      '

# Right align
print("{:>10}".format("right"))   # '     right'

# Center align
print("{:^10}".format("center"))  # '  center  '

# Padding with characters
print("{:*^10}".format("star"))   # '***star***'
```

### Number Representations

```python
# Binary
print("{:b}".format(10))  # 1010

# Hexadecimal
print("{:x}".format(255))  # ff

# Octal
print("{:o}".format(8))    # 10
```

## Advanced Formatting

### Type Conversion

```python
# String representation
print("{!s}".format(123))  # '123'

# repr() representation
print("{!r}".format("text"))  # "'text'"
```

### Dynamic Formatting

```python
# Format specifications can be variables
width = 10
precision = 2
value = 3.14159
print("{:{width}.{prec}f}".format(value, width=width, prec=precision))
# '      3.14'
```

### Date Formatting

```python
from datetime import datetime
print("{:%Y-%m-%d}".format(datetime.now()))
# e.g., '2023-05-15'
```

## Comparison with Other Formatting Methods

| Method       | Python Version | Features                            |
| ------------ | -------------- | ----------------------------------- |
| `%` operator | All            | Basic, limited features             |
| `format()`   | 2.6+           | Flexible, powerful                  |
| f-strings    | 3.6+           | Most readable, evaluated at runtime |

## Best Practices

1. **Use f-strings in Python 3.6+** for better readability:

   ```python
   name = "Alice"
   print(f"Hello, {name}")  # Preferred in modern Python
   ```

2. **Use `format()` when:**

   - You need compatibility with Python 2.7 or 3.5
   - Format strings are defined separately from values
   - You need advanced formatting features

3. **For localization**, consider template strings or localization libraries.

## Real-World Examples

### 1. Table Formatting

```python
data = [("Alice", 28, "Engineer"),
        ("Bob", 32, "Designer")]

for name, age, job in data:
    print("{:<10} | {:>3} | {:<15}".format(name, age, job))
```

### 2. Complex Number Formatting

```python
for num in [1234.5678, -9876.5432, 3.14159]:
    print("{:>15,.2f}".format(num))
# Output:
#       1,234.57
#      -9,876.54
#           3.14
```

### 3. Custom String Templates

```python
template = "Welcome, {name}! Your ID is {id:0>6}"
print(template.format(name="Charlie", id=42))
# Welcome, Charlie! Your ID is 000042
```

The `format()` method remains a powerful tool for string formatting in Python, especially when you need more control than f-strings provide or when working with older Python versions.


The `format()` method in Python is a powerful way to format strings. It allows you to insert values into a string using curly braces `{}` as placeholders. Here’s a comprehensive look at how to use it, along with examples.

### Basic Syntax

```python
string.format(value1, value2, ...)
```

### Placeholder Syntax

In the string, you can define placeholders using curly braces `{}`. You can also specify the order or use named parameters.

### Examples

#### 1. Basic Usage

```python
name = "Alice"
age = 30
formatted_string = "My name is {} and I am {} years old.".format(name, age)
print(formatted_string)  # Output: My name is Alice and I am 30 years old.
```

#### 2. Positional Arguments

You can use positional indexing within the placeholders.

```python
formatted_string = "My name is {0} and I am {1} years old. {0} loves Python.".format(name, age)
print(formatted_string)  # Output: My name is Alice and I am 30 years old. Alice loves Python.
```

#### 3. Named Arguments

You can also use named arguments, which makes your code more readable.

```python
formatted_string = "My name is {name} and I am {age} years old.".format(name=name, age=age)
print(formatted_string)  # Output: My name is Alice and I am 30 years old.
```

#### 4. Format Specifiers

You can specify formats for numbers, dates, etc.

```python
pi = 3.14159
formatted_string = "Pi is approximately {:.2f}.".format(pi)
print(formatted_string)  # Output: Pi is approximately 3.14.
```

#### 5. Padding and Alignment

You can control padding and alignment using the format specifiers.

```python
formatted_string = "{:<10} | {:^10} | {:>10}".format("Left", "Center", "Right")
print(formatted_string)
# Output:
# Left      |   Center   |      Right
```

#### 6. Formatting Dictionaries

You can unpack dictionaries using `**`.

```python
data = {'name': 'Alice', 'age': 30}
formatted_string = "My name is {name} and I am {age} years old.".format(**data)
print(formatted_string)  # Output: My name is Alice and I am 30 years old.
```

### Conclusion

The `format()` method is versatile for string formatting in Python, making it easier to create dynamic strings. While Python 3.6 and later introduced f-strings (which are often easier to read), knowing how to use `format()` is valuable, especially in older codebases.


Alright, let's dive deep into the `format()` method in Python strings. It's a powerful and versatile way to create formatted strings by embedding expressions inside string literals.

**What `format()` Does:**

The `format()` method allows you to perform string formatting. It works by taking positional and keyword arguments, and these arguments are substituted into the string at the placeholders defined by curly braces `{}`.

**Syntax:**

There are two main ways to use placeholders within the string literal:

1.  **Implicit Positional Formatting:** Empty curly braces `{}` are replaced by arguments passed to `format()` in the order they are provided.

    ```python
    "String with {} and {}".format(value1, value2)
    ```

2.  **Explicit Positional Formatting:** Curly braces contain integers (starting from 0) that refer to the index of the argument passed to `format()`.

    ```python
    "String with {0}, {1}, and again {0}".format(value1, value2)
    ```

3.  **Keyword Argument Formatting:** Curly braces contain names that correspond to keyword arguments passed to `format()`.

    ```python
    "String with {name} and {count}".format(name="Alice", count=10)
    ```

4.  **Attribute Access:** You can access attributes of objects passed to `format()` within the curly braces.

    ```python
    class Point:
        def __init__(self, x, y):
            self.x = x
            self.y = y

    point = Point(3, 5)
    "Point's x: {0.x}, y: {0.y}".format(point)
    ```

5.  **Item Access:** You can access items of sequences (like lists and tuples) or dictionaries using their index or key within the curly braces.

    ```python
    my_list = [10, 20]
    "List item 0: {0[0]}, item 1: {0[1]}".format(my_list)

    my_dict = {"a": 100, "b": 200}
    "Dict value for 'a': {0[a]}, value for 'b': {0[b]}".format(my_dict)
    ```

**Format Specifiers:**

The power of `format()` lies in its format specifiers, which allow you to control how values are presented. These specifiers follow a colon `:` after the field name (or index). The general format of a specifier is:

```
{[field_name]:[fill][align][sign][#][0][width][grouping_option][.precision][type]}
```

Let's break down the components:

- **`field_name`:** (Optional) The index or keyword of the argument to be formatted.
- **`:`:** Separates the field name from the format specifier.
- **`fill`:** (Optional) A character used to pad the value when it needs to be aligned to a certain width. Can be any character except `{}`.
- **`align`:** (Optional) Specifies the alignment of the value within the field width:
  - `<`: Left-aligned (default for strings).
  - `>`: Right-aligned (default for numbers).
  - `^`: Centered.
  - `=`: Forces the sign (+ or -) to be to the leftmost position, and pads the remaining digits. Only valid for numeric types.
- **`sign`:** (Optional) Controls the display of signs for numeric types:
  - `+`: Show the sign for both positive and negative numbers.
  - `-`: Show the sign only for negative numbers (default).
  - ` `: Use a leading space for positive numbers and a minus sign for negative numbers.
- **`#`:** (Optional) For integer types, adds a prefix:
  - `0b` for binary.
  - `0o` for octal.
  - `0x` for hexadecimal.
    For float, complex, and decimal types, it forces the output to always contain a decimal point, even if no digits follow it, and for floats, it also prevents trailing zeros from being removed.
- **`0`:** (Optional) For numeric types, pads the number with leading zeros to the specified width. Equivalent to `fill='0'` and `align='='`.
- **`width`:** (Optional) An integer specifying the minimum width of the field. The value will be padded if it's shorter than this width.
- **`grouping_option`:** (Optional) Used for numeric types to add thousands separators:
  - `,`: Uses a comma as the thousands separator.
  - `_`: Uses an underscore as the thousands separator (available in Python 3.6+).
- **`.precision`:** (Optional) For floating-point numbers, specifies the number of digits to display after the decimal point. For non-numeric types, it indicates the maximum field width.
- **`type`:** (Optional) Specifies the type of formatting to be applied to the value:
  - **Integer types:**
    - `b`: Binary format.
    - `c`: Character corresponding to the Unicode code point.
    - `d`: Decimal integer (default).
    - `o`: Octal format.
    - `x`: Hexadecimal format (lowercase).
    - `X`: Hexadecimal format (uppercase).
    - `n`: Number format. Same as 'd', but uses the current locale's thousands separator.
  - **Floating-point and related types:**
    - `e`: Scientific notation (lowercase 'e').
    - `E`: Scientific notation (uppercase 'E').
    - `f`: Fixed-point notation (lowercase).
    - `F`: Fixed-point notation (uppercase).
    - `g`: General format. Switches between fixed-point and scientific notation.
    - `G`: General format. Switches between fixed-point and scientific notation (uppercase 'E').
    - `n`: Number format. Same as 'g', but uses the current locale's thousands separator.
    - `%`: Percentage. Multiplies the number by 100 and displays with a trailing '%'.
  - **String types:**
    - `s`: String format (default).

**Code Examples:**

Let's see various formatting options in action:

```python
name = "Alice"
age = 30
price = 19.99
count = 1000000

# Implicit positional formatting
print("Name: {}, Age: {}".format(name, age))
# Output: Name: Alice, Age: 30

# Explicit positional formatting
print("Age: {1}, Name: {0}".format(name, age))
# Output: Age: 30, Name: Alice

# Keyword argument formatting
print("Name: {person}, Age: {years}".format(person=name, years=age))
# Output: Name: Alice, Age: 30

# Basic alignment
print("{:<10}".format(name))   # Left-aligned, width 10
print("{:>10}".format(age))    # Right-aligned, width 10
print("{:^10}".format(price))  # Centered, width 10
# Output (with spaces):
# Alice
#         30
#   19.99

# Fill character
print("{:*<10}".format(name))
print("{:0>10}".format(age))
print("{:-^10}".format(price))
# Output:
# Alice*****
# 0000000030
# ---19.99---

# Sign
print("{:+d}".format(age))
print("{:-d}".format(-age))
print("{: d}".format(age))
print("{: d}".format(-age))
# Output:
# +30
# -30
#  30
# -30

# Thousands separator
print("{:,}".format(count))
print("{:_}".format(count))
# Output:
# 1,000,000
# 1_000_000

# Precision for floats
print("{:.2f}".format(price))
print("{:.1%}".format(0.856))
# Output:
# 19.99
# 85.6%

# Type specifiers for integers
number = 25
print("{:b}".format(number))  # Binary
print("{:o}".format(number))  # Octal
print("{:x}".format(number))  # Hexadecimal (lowercase)
print("{:X}".format(number))  # Hexadecimal (uppercase)
print("{:c}".format(65))     # Character
# Output:
# 11001
# 31
# 19
# 19
# A

# Type specifiers for floats
value = 123.456789
print("{:e}".format(value))  # Scientific notation (lowercase)
print("{:E}".format(value))  # Scientific notation (uppercase)
print("{:f}".format(value))  # Fixed-point
print("{:.2g}".format(value)) # General format (2 significant digits)
# Output:
# 1.234568e+02
# 1.234568E+02
# 123.456789
# 1.2e+02

# Combining format specifiers
print("{:*^15,.2f}".format(price * count / 100))
# Output: ***199,900.00***
```

**f-strings (Formatted String Literals):**

Introduced in Python 3.6, f-strings provide an even more concise and readable way to embed expressions. They use an `f` or `F` prefix before the opening quote, and you can directly include expressions inside the curly braces. Format specifiers work the same way within f-strings.

```python
name = "Bob"
age = 25
print(f"Name: {name}, Age: {age}")
print(f"{name:<10} is {age:>3} years old.")
print(f"The price is {price:.2f}")
```

**Conclusion:**

The `format()` method (and f-strings) in Python offer a powerful and flexible way to create formatted strings. By understanding the different ways to use placeholders and the various format specifiers, you can precisely control the presentation of your data in strings, making your output more readable and professional. F-strings are generally preferred for their conciseness, but `format()` remains valuable, especially in older Python versions or when you need to build format strings dynamically.
